src/main/java/com/streamr/client/protocol/control_layer/ControlMessageAdapter.java

```diff
@@ -0,0 +1,57 @@
+package com.streamr.client.protocol.control_layer;
+
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.JsonReader;
+import com.squareup.moshi.JsonWriter;
+import com.streamr.client.exceptions.UnsupportedMessageException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.util.HashMap;
+
+public class ControlMessageAdapter extends JsonAdapter<ControlMessage> {
+    private static final HashMap<Integer, ControlLayerAdapter<? extends ControlMessage>> adapters = new HashMap<>();
```

Comment:
I see, maybe you created `ControlLayerAdapter` for this to work. But couldn't you have 
`HashMap<Integer, JsonAdapter<? extends ControlMessage>>` just as well?